zcash_client_backend/src/data_api/ll.rs

```diff
@@ -0,0 +1,554 @@
+        self.sapling_bundle().into_iter().flat_map(|bundle| {
+            bundle
+                .shielded_spends()
+                .iter()
+                .map(|spend| spend.nullifier())
+        })
+    }
+
+    #[cfg(feature = "orchard")]
+    fn orchard_spent_note_nullifiers(&self) -> impl Iterator<Item = &::orchard::note::Nullifier> {
+        self.orchard_bundle()
+            .into_iter()
+            .flat_map(|bundle| bundle.actions().iter().map(|action| action.nullifier()))
+    }
+
+    fn fee_paid<E, F>(&self, get_prevout: F) -> Result<Option<Zatoshis>, E>
+    where
+        E: From<BalanceError>,
+        F: FnMut(&OutPoint) -> Result<Option<Zatoshis>, E>,
+    {
+        self.data().fee_paid(get_prevout)
+    }
+}
+
+/// A capability trait that provides low-level wallet database read operations. These operations
+/// are used to provide standard implementations for certain [`WalletWrite`] trait methods.
+///
+/// [`WalletWrite`]: super::WalletWrite
+pub trait LowLevelWalletRead {
+    /// The type of errors that may be generated when querying a wallet data store.
+    type Error: Debug;
+
+    /// The type of the account identifier.
+    ///
+    /// An account identifier corresponds to at most a single unified spending key's worth of spend
+    /// authority, such that both received notes and change spendable by that spending authority
+    /// will be interpreted as belonging to that account.
+    type AccountId: Debug + Copy + Eq + Hash;
+
+    /// A wallet-internal transaction identifier.
+    type TxRef: Copy + Eq;
+
+    /// Returns the set of account identifiers for accounts that spent notes and/or UTXOs in the
+    /// construction of the given transaction.
+    fn get_funding_accounts<T: TxMeta>(
+        &self,
+        tx: &T,
+    ) -> Result<HashSet<Self::AccountId>, Self::Error> {
+        let mut funding_accounts = HashSet::new();
+
+        #[cfg(feature = "transparent-inputs")]
+        funding_accounts.extend(self.detect_accounts_transparent(tx.transparent_spends())?);
+
+        funding_accounts.extend(self.detect_accounts_sapling(tx.sapling_spent_note_nullifiers())?);
+
+        #[cfg(feature = "orchard")]
+        funding_accounts.extend(self.detect_accounts_orchard(tx.orchard_spent_note_nullifiers())?);
+
+        Ok(funding_accounts)
+    }
+
+    /// Returns the most likely wallet address that corresponds to the protocol-level receiver of a
+    /// note or UTXO.
+    ///
+    /// If the wallet database has stored a wallet address that contains the given receiver, then
+    /// that address is returned; otherwise, a new address containing that receiver is generated.
+    fn select_receiving_address(
+        &self,
+        account: Self::AccountId,
+        receiver: &Receiver,
+    ) -> Result<Option<ZcashAddress>, Self::Error>;
+
+    /// Detects and returns the identifier for the account that generated the given address, if the
+    /// address belongs to an account in the wallet.
```

Comment:
"generated" is the wrong word here, given that standalone transparent addresses are now supported in backends. Maybe "the account to which the given address belongs, if any"?